caddyhttp/templates/templates.go

```diff
@@ -4,90 +4,92 @@ package templates
+		// check if extension matches
 		reqExt := path.Ext(fpath)
-
 		for _, ext := range rule.Extensions {
 			if reqExt == ext {
-				// Create execution context
-				ctx := httpserver.NewContextWithHeader(w.Header())
-				ctx.Root = t.FileSys
-				ctx.Req = r
-				ctx.URL = r.URL
+				// get a buffer from the pool and make a response recorder
+				buf := t.BufPool.Get().(*bytes.Buffer)
+				buf.Reset()
+				defer t.BufPool.Put(buf)
+				respRec := &httptest.ResponseRecorder{
+					HeaderMap: make(http.Header),
+					Body:      buf,
+				}
+
+				// pass request up the chain to let another middleware provide us the template
+				code, err := t.Next.ServeHTTP(respRec, r)
+				if code >= 300 {
+					return code, err
+				}
 
-				// New template
+				// copy the buffered response headers to the real response writer
+				for field, val := range respRec.HeaderMap {
+					w.Header()[field] = val
+				}
+
+				// create a new template
 				templateName := filepath.Base(fpath)
 				tpl := template.New(templateName)
 
-				// Set delims
+				// set delimiters
 				if rule.Delims != [2]string{} {
 					tpl.Delims(rule.Delims[0], rule.Delims[1])
 				}
 
-				// Add custom functions
+				// add custom functions
 				tpl.Funcs(httpserver.TemplateFuncs)
 
-				// Build the template
-				templatePath := filepath.Join(t.Root, fpath)
-				tpl, err := tpl.ParseFiles(templatePath)
+				// parse the template
+				parsedTpl, err := tpl.Parse(respRec.Body.String())
 				if err != nil {
-					if os.IsNotExist(err) {
-						return http.StatusNotFound, nil
-					} else if os.IsPermission(err) {
-						return http.StatusForbidden, nil
-					}
 					return http.StatusInternalServerError, err
 				}
 
-				// Execute it
-				buf := t.BufPool.Get().(*bytes.Buffer)
+				// create execution context for the template template
+				ctx := httpserver.NewContextWithHeader(w.Header())
+				ctx.Root = t.FileSys
+				ctx.Req = r
+				ctx.URL = r.URL
+
+				// execute the template
 				buf.Reset()
-				defer t.BufPool.Put(buf)
-				err = tpl.Execute(buf, ctx)
+				err = parsedTpl.Execute(buf, ctx)
 				if err != nil {
 					return http.StatusInternalServerError, err
 				}
 
-				// If Content-Type isn't set here, http.ResponseWriter.Write
-				// will set it according to response body. But other middleware
-				// such as gzip can modify response body, then Content-Type
-				// detected by http.ResponseWriter.Write is wrong.
-				ctype := mime.TypeByExtension(ext)
-				if ctype == "" {
-					ctype = http.DetectContentType(buf.Bytes())
-				}
-				w.Header().Set("Content-Type", ctype)
+				modTime, _ := time.Parse(http.TimeFormat, w.Header().Get("Last-Modified"))
 
-				templateInfo, err := os.Stat(templatePath)
-				if err == nil {
-					// add the Last-Modified header if we were able to read the stamp
-					httpserver.SetLastModifiedHeader(w, templateInfo.ModTime())
-				}
-				buf.WriteTo(w)
+				http.ServeContent(w, r, templateName, modTime, bytes.NewReader(buf.Bytes()))
```

Comment:
> // TODO: This performs ~15% worse (req/s) than the old method; why?

I did some investigation find that the culprit is `http.ServeContent`. It will use `io.Copy` to write content to the `w` which allocates a lot of memory. For now, there are two places using `http.ServeContent`, one is here and the other one is `staticfiles` middleware.

I take a memory profile as reference:

```
         .          .     19:func (t Templates) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
         .          .     20:   // TODO: This performs ~15% worse (req/s) than the old method; why?
         .          .     21:   // TODO: Need to somehow embed upcoming httpserver.ResponseWriterWrapper
         .          .     22:   // into the ResponseRecorder, so it implements the needed interfaces
         .          .     23:
         .          .     24:   for _, rule := range t.Rules {
         .          .     25:           if !httpserver.Path(r.URL.Path).Matches(rule.Path) {
         .          .     26:                   continue
         .          .     27:           }
         .          .     28:
         .          .     29:           // check for index files
         .          .     30:           fpath := r.URL.Path
         .          .     31:           if idx, ok := httpserver.IndexFile(t.FileSys, fpath, rule.IndexFiles); ok {
         .          .     32:                   fpath = idx
         .          .     33:           }
         .          .     34:
         .          .     35:           // check if extension matches
         .          .     36:           reqExt := path.Ext(fpath)
         .          .     37:           for _, ext := range rule.Extensions {
         .          .     38:                   if reqExt == ext {
         .          .     39:                           // get a buffer from the pool and make a response recorder
         .          .     40:                           buf := t.BufPool.Get().(*bytes.Buffer)
         .          .     41:                           buf.Reset()
         .          .     42:                           defer t.BufPool.Put(buf)
         .          .     43:                           respRec := &httptest.ResponseRecorder{
       4MB        4MB     44:                                   HeaderMap: make(http.Header),
       2MB        2MB     45:                                   Body:      buf,
         .          .     46:                           }
         .          .     47:
         .          .     48:                           // pass request up the chain to let another middleware provide us the template
         .     1.29GB     49:                           code, err := t.Next.ServeHTTP(respRec, r)
         .          .     50:                           if code >= 300 {
         .          .     51:                                   return code, err
         .          .     52:                           }
         .          .     53:
         .          .     54:                           // copy the buffered response headers to the real response writer
         .          .     55:                           for field, val := range respRec.HeaderMap {
         .          .     56:                                   w.Header()[field] = val
         .          .     57:                           }
         .          .     58:
         .          .     59:                           // create a new template
         .          .     60:                           templateName := filepath.Base(fpath)
         .          .     61:                           tpl := template.New(templateName)
         .          .     62:
         .          .     63:                           // set delimiters
         .          .     64:                           if rule.Delims != [2]string{} {
         .          .     65:                                   tpl.Delims(rule.Delims[0], rule.Delims[1])
         .          .     66:                           }
         .          .     67:
         .          .     68:                           // add custom functions
         .          .     69:                           tpl.Funcs(httpserver.TemplateFuncs)
         .          .     70:
         .          .     71:                           // parse the template
         .    67.01MB     72:                           parsedTpl, err := tpl.Parse(respRec.Body.String())
         .          .     73:                           if err != nil {
         .          .     74:                                   return http.StatusInternalServerError, err
         .          .     75:                           }
         .          .     76:
         .          .     77:                           // create execution context for the template template
         .          .     78:                           ctx := httpserver.NewContextWithHeader(w.Header())
         .          .     79:                           ctx.Root = t.FileSys
         .          .     80:                           ctx.Req = r
         .          .     81:                           ctx.URL = r.URL
         .          .     82:
         .          .     83:                           // execute the template
         .          .     84:                           buf.Reset()
    2.50MB   179.11MB     85:                           err = parsedTpl.Execute(buf, ctx)
         .          .     86:                           if err != nil {
         .          .     87:                                   return http.StatusInternalServerError, err
         .          .     88:                           }
         .          .     89:
         .          .     90:                           modTime, _ := time.Parse(http.TimeFormat, w.Header().Get("Last-Modified"))
         .          .     91:
         .          .     92:                           http.ServeContent(w, r, templateName, modTime, bytes.NewReader(buf.Bytes()))
         .          .     93:
         .     1.22GB     94:                           return http.StatusOK, nil
         .          .     95:                   }
         .          .     96:           }
         .          .     97:   }
```